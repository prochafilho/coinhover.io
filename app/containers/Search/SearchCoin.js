import React from 'react';
import { connect } from 'react-redux';
import { findCoins } from '../../services/coinFactory';
import { addCoin, setSearch } from '../../actions';

class SearchCoin extends React.Component {
  constructor(props) {
    super(props);

    console.log('SearchCoin', props);

    this.state = {
      coins: props.coins,
      saved: props.search[0]
    };

    this.close = this.close.bind(this);
    this.clickCoin = this.clickCoin.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    this.coinInput.focus();
  }

  handleChange() {
    const text = document.getElementById('coin-search').value;

    const search = (txt) => {
      const searchedCoins = findCoins(txt);
      this.props.setSearch(searchedCoins);
      this.setState({ coins: searchedCoins });
    };

    const clearSearch = () => {
      this.props.setSearch([]);
      this.setState({ coins: this.state.saved });
    };

    const handleUpdate = num => (num > 1 ? search(text) : clearSearch());

    return handleUpdate();
  }

  clickCoin(coin) {
    this.props.addCoin(coin);
    this.props.closeSearch();
  }

  close() {
    this.props.closeSearch();
  }

  render() {
    const { coins } = this.props;
    console.log('coins.collection', coins.collection);

    const searchCoins = coins.collection.map(coin => (
      <li key={coin.id}>
        <button onClick={() => this.clickCoin(coin)}>
          <span>{ coin.name }</span>
        </button>
      </li>
    ));

    return (
      <div id="search-coin-component">
        <input
          type="text"
          id="coin-search"
          className="coin-search-input fl"
          placeholder="Search"
          onChange={() => this.handleChange()}
          ref={(input) => { this.coinInput = input; }}
        />
        <div className="cancel-search">
          <div className="close-x" />
        </div>
        <div className="coins-container">
          <div className="coin-select">
            <ul>
              { searchCoins }
            </ul>
          </div>
        </div>
        <div className="overlay" role="button" tabIndex={0} onClick={this.close} />
      </div>
    );
  }
}

const mapStateToProps = ({ coins, search }) => ({
  coins,
  search
});

const mapDispatchToProps = dispatch => ({
  addCoin: (...args) => { dispatch(addCoin(...args)); },
  setSearch: (...args) => { dispatch(setSearch(...args)); }
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchCoin);
