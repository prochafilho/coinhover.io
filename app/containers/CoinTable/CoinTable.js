/* eslint-disable import/no-named-as-default */
import React from 'react';
import { connect } from 'react-redux';

import CoinThead from '../../components/CoinThead/CoinThead';
import CoinRow from '../CoinRow/CoinRow';
import SearchCoin from '../Search/SearchCoin';

export class CoinTable extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      search: false
    };

    this.closeSearch = this.closeSearch.bind(this);
    console.log('CoinTable', props);
  }

  closeSearch() {
    this.setState({ search: false });
  }

  render() {
    const assets = this.props.portfolio ? this.props.portfolio : [];
    // const { assets } = this.props ? this.props : [];
    console.log('CoinTable render assets', assets);

    const handleClick = (e) => {
      e.preventDefault();
      this.setState({ search: true });
    };

    return (
      <section className="asset-table">
        <CoinThead />
        { assets.map(asset =>
          (<CoinRow
            key={asset.id}
            asset={asset}
            balance={asset.balance}
            price_usd={asset.price_usd}
            percentage={asset.percentage}
          />)) }
        <div className="add-asset-row">
          { this.state.search ? (
            <SearchCoin closeSearch={this.closeSearch} />
          ) : (
            <div className="add-btn" role="button" tabIndex={0} onClick={handleClick}>
              + Add Asset
            </div>
          )}
        </div>
      </section>
    );
  }
}

const mapStateToProps = state => (state);

const CoinTableContainer = CoinTable;
export default connect(mapStateToProps, null)(CoinTableContainer);
