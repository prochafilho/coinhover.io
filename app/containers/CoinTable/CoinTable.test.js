import React from 'react';
import * as enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import { CoinTable } from './CoinTable';
import CoinThead from '../../components/CoinThead/CoinThead';
import SearchCoin from '../Search/SearchCoin';

const coinTable = enzyme.shallow(<CoinTable />);

describe('<CoinTable /> component', () => {
  it('should render', () => {
    const tree = toJson(coinTable);
    expect(tree).toMatchSnapshot();
  });

  it('contains the CoinThead component', () => {
    expect(coinTable.find(CoinThead).length).toBe(1);
  });

  it('Clicking the Add Asset button should enable the searchCoin component', () => {
    coinTable.setState({ search: true });
    expect(coinTable.find(SearchCoin).length).toBe(1);
  });
});
