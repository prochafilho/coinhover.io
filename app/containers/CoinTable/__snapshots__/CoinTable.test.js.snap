// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoinTable /> component should render 1`] = `
<section
  className="asset-table"
>
  <CoinThead />
  <div
    className="add-asset-row"
  >
    <div
      className="add-btn"
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      + Add Asset
    </div>
  </div>
</section>
`;
